pkg/lib/kitfile/generate.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2024 The KitOps Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package kitfile
+
+import (
+	"fmt"
+	"io/fs"
+	"kitops/pkg/artifact"
+	"kitops/pkg/output"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/licensecheck"
+)
+
+var modelWeightsSuffixes = []string{
+	".safetensors", ".pkl",
+	// Pytorch suffixes
+	".bin", ".pth", ".pt", ".mar", ".pt2", ".ptl",
+	// Tensorflow
+	".pb", ".ckpt", ".tflite", ".tfrecords",
+	// NumPy
+	".npy", ".npz",
+	// Keras and others
+	".keras", ".h5", ".caffemodel", ".pmml", ".coreml",
+	// Other suffixes
+	".gguf", ".ggml", ".ggmf", ".llamafile", ".onnx",
+}
+
+var docsSuffixes = []string{
+	".md", ".adoc", ".html", ".pdf",
+}
+
+var metadataSuffixes = []string{
+	".json", ".yaml", ".xml", ".csv", ".txt",
+}
+
+var datasetSuffixes = []string{
+	".tar", ".zip",
+}
+
+// Generate a basic Kitfile by looking at the contents of a directory. Parameter
+// packageOpt can be used to define metadata for the Kitfile (i.e. the package
+// section), which is left empty if the parameter is nil.
+func GenerateKitfile(baseDir string, packageOpt *artifact.Package) (*artifact.KitFile, error) {
```

Comment:
I think this needs to look into the sub directories as well. If somebody organizes their files under `data/` for instance `data/training.csv` it should be detected.

Reply:
Yeah, I was hoping to defer that to a future improvement since it introduces a fair bit of complexity in the general case.

* If we're trying to grab specific files (e.g. a dataset layer for `path: data/training.csv`, then we need to decide what to do with the rest of the directory
  * If that's the only file, we don't need to include the directory itself, since it'll be included in the dataset layer, which means signalling another condition back to the caller
  * If there are other files, we need to either classify them or include the whole directory as code (?) layer
  * Then we also need to handle the case where we've got a catch-all `code` section
* If we want to include the whole directory (i.e. `path: data/`) then we need to figure out a heuristic for deciding what type the whole directory is (most common file type? I'm not sure here)
* If a `data/` dir has 20 .csv files, do we include a single `data/` layer, or generate a Kitfile with 20 dataset layers (one for each file).

Reply:
Alright, I've added a (messy) implementation of this in commit adf3372. For e.g. https://huggingface.co/onnx-community/dpt-dinov2-small-kitti/tree/main it generates the following Kitfile:

```yaml
manifestVersion: 1.0.0
model:
  name: onnx/model
  path: onnx/model.onnx
  parts:
    - path: onnx/model_bnb4.onnx
    - path: onnx/model_fp16.onnx
    - path: onnx/model_int8.onnx
    - path: onnx/model_q4.onnx
    - path: onnx/model_q4f16.onnx
    - path: onnx/model_quantized.onnx
    - path: onnx/model_uint8.onnx
    - path: config.json
    - path: preprocessor_config.json
    - path: quantize_config.json
code:
  - path: .
docs:
  - path: README.md
    description: Readme file
```
where the `code` section includes the `.gitattributes` file and the `.git` directory (since I cloned it locally)